modules/jetty-http-server/src/main/java/com/spotify/apollo/http/server/ApolloRequestHandler.java

```diff
@@ -121,24 +112,33 @@ Request asApolloRequest(HttpServletRequest req) throws IOException {
 
     final ImmutableMap<String, String> headers = headersBuilder.build();
 
-    Request result = Request.forUri(uri, method)
-        .withHeaders(headers);
+    Request request = Request.forUri(uri, method).withHeaders(headers);
 
-    final Optional<String> callingService = result.header("X-Calling-Service");
+    final Optional<String> callingService = request.header("X-Calling-Service");
     if (callingService.isPresent() && !callingService.get().isEmpty()) {
-      result = result.withService(callingService.get());
+      request = request.withService(callingService.get());
     }
 
+    Optional<ByteString> payload = readPayload(req, contentLength);
+
     if (payload.isPresent()) {
-      result = result.withPayload(payload.get());
+      request = request.withPayload(payload.get());
     }
 
-    return result;
+    return request;
   }
 
-  private ByteString readPayload(HttpServletRequest req, int contentLength) throws IOException {
+  private Optional<ByteString> readPayload(HttpServletRequest req, int contentLength) throws IOException {
     final InputStream input = new BufferedInputStream(req.getInputStream());
-    return ByteString.read(input, contentLength);
+    if (contentLength < 0) {
+      // contentLength = -1 may be returned when using Transfer-Encoding: chunked
+      // (RFC 7230, section 3.3.1: Transfer-Encoding) even though data is being transferred in a series of chunks.
+      // In that case try read this data before concluding that there is no payload.
+      final ByteString byteString = ByteString.of(ByteStreams.toByteArray(input));
+      return byteString.size() == 0 ? Optional.empty() : Optional.of(byteString);
```

Comment:
On reading this PR, I notice that apollo treats content-length = 0 as 'payload present but empty'. It looks to me like this code means that the case of (chunked + content-length = -1 + content actually has zero length) leads to different behaviour. Is there a way of ensuring that the behaviour would be identical in that case? There doesn't seem to be any unit tests guarding the old behaviour either. And it's unlikely to be a very important distinction I guess, so I'm fine with deciding to not fix that.

Reply:
Hi! I currently don't have much time to look into this. I can however say that we've been running a patched version (with the code on this PR) of Apollo for a few months now and have had no problem. I think I could take a further look into this at some point when I'm less busy, but I agree it would be fine to merge this as is.